File: src/sdk/cloud-py/ratel_ai_cloud/skill_sync.py
"""`SkillSync` — pull-syncs a source's published skills into a host-owned
`SkillCatalog` under the ownership rule: the loader mutates only the ids it
synced itself; a host-registered skill with a colliding id is reported in
`SyncResult.conflicts` and never touched. The replica is in-memory only;
staleness surfaces as `last_synced_at` / `consecutive_failures` (ADR-0003).
"""

from __future__ import annotations

import random
import threading
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, Optional

from ratel_ai import Skill, SkillCatalog, SourceConfig

from .canonical import SKILL_FIELDS
from .errors import ApiError, AuthError
from .fetch_catalog import CatalogChanged, FetchCatalogResult, fetch_catalog

__all__ = ["SkillSync", "SyncResult", "skills_equal"]

FetchImpl = Callable[[SourceConfig, Optional[str]], FetchCatalogResult]


@dataclass(frozen=True)
class SyncResult:
    added: int
    updated: int
    removed: int
    # Ids present in the catalog but not owned by this sync — host skills the
    # ownership rule refused to overwrite.
    conflicts: list[str]
    # True when the source answered 304 for the cached etag.
    unchanged: bool


def skills_equal(a: Skill, b: Skill) -> bool:
    """Field-wise equality over exactly the 7 wire fields — the same set the
    frozen ETag projection hashes, so an identical resync emits zero churn."""
    return all(getattr(a, f) == getattr(b, f) for f in SKILL_FIELDS)


def _to_skill(wire: dict[str, Any]) -> Skill:
    """Project a wire skill to the 7 fields (unknown fields are ignored)."""
    return Skill(
        id=wire["id"],
        name=wire["name"],
        description=wire["description"],
        tags=list(wire["tags"]),
        tools=list(wire["tools"]),
        metadata={k: list(v) for k, v in wire["metadata"].items()},
        body=wire["body"],
    )


def _jittered_interval(interval_s: float) -> float:
    """The next tick's delay: the interval with ±10% jitter, so a fleet of
    loaders never revalidates in lockstep."""
    return interval_s * (0.9 + 0.2 * random.random())


@dataclass
class SkillSync:
    """A sync handle for one `(source url, scope)`. `refresh()` is one
    conditional pull + diff apply; `start()` keeps refreshing on a jittered
    timer chain until `stop()` or a permanent auth failure."""

    catalog: SkillCatalog
    config: SourceConfig
    fetch_impl: FetchImpl = field(default=fetch_catalog)

    def __post_init__(self) -> None:
        self._refresh_lock = threading.Lock()
        self._state_lock = threading.Lock()
        self._etag: str | None = None
        self._synced_ids: set[str] = set()
        self._last_synced_at: datetime | None = None
        self._consecutive_failures = 0
        self._interval_s: float | None = None
        self._timer: threading.Timer | None = None
        self._stopped = False
        self._auth_failed = False

    @property
    def last_synced_at(self) -> datetime | None:
        return self._last_synced_at

    @property
    def consecutive_failures(self) -> int:
        return self._consecutive_failures

    @property
    def owned_count(self) -> int:
        return len(self._synced_ids)

    def refresh(self) -> SyncResult:
        """One conditional pull + diff apply, serialized via a lock. Raises the
        typed fetch errors; a failure keeps the replica and bumps
        `consecutive_failures`."""
        with self._refresh_lock:
            try:
                result = self.fetch_impl(self.config, self._etag)
            except Exception:
                self._consecutive_failures += 1
                raise
            sync_result = (
                self._apply(result)
                if isinstance(result, CatalogChanged)
                else SyncResult(added=0, updated=0, removed=0, conflicts=[], unchanged=True)
            )
            self._last_synced_at = datetime.now(timezone.utc)
            self._consecutive_failures = 0
            return sync_result

    def _apply(self, result: CatalogChanged) -> SyncResult:
        added = updated = removed = 0
        conflicts: list[str] = []
        wire_ids: set[str] = set()
        for wire in result.catalog.skills:
            skill = _to_skill(wire)
            wire_ids.add(skill.id)
            if skill.id in self._synced_ids:
                existing = self.catalog.get(skill.id)
                if existing is None or not skills_equal(existing, skill):
                    self.catalog.upsert(skill)
                    updated += 1
            elif self.catalog.has(skill.id):
                conflicts.append(skill.id)
            else:
                self.catalog.register(skill)
                self._synced_ids.add(skill.id)
                added += 1
        for gone in sorted(self._synced_ids - wire_ids):
            self.catalog.remove(gone)
            self._synced_ids.discard(gone)
            removed += 1
        self._etag = result.etag
        return SyncResult(
            added=added, updated=updated, removed=removed, conflicts=conflicts, unchanged=False
        )

    def start(self, interval_s: float) -> None:
        """Keep refreshing every `interval_s` seconds (±10% jitter) on a daemon
        timer chain. Transient errors keep the chain; an auth failure stops it
        permanently."""
        with self._state_lock:
            if self._auth_failed:
                return
            self._stopped = False
            self._interval_s = interval_s
            self._schedule_locked()

    def stop(self) -> None:
        """Cancel the timer chain. Idempotent and safe to race with a tick."""
        with self._state_lock:
            self._stopped = True
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None

    def _schedule_locked(self) -> None:
        if self._stopped or self._auth_failed or self._interval_s is None:
            return
        timer = threading.Timer(_jittered_interval(self._interval_s), self._tick)
        timer.daemon = True
        self._timer = timer
        timer.start()

    def _tick(self) -> None:
        try:
            self.refresh()
        except AuthError:
            with self._state_lock:
                self._auth_failed = True
                self._timer = None
            return
        except ApiError:
            pass  # transient — the replica stays live, the chain continues
        with self._state_lock:
            self._timer = None
            self._schedule_locked()

File: src/sdk/cloud/src/skill-sync.ts
/**
 * Pull-sync of a catalog source into a local `SkillCatalog`. One conditional
 * fetch per refresh; the wire diff is applied under the ownership rule: the
 * sync only ever mutates skills it created, host-registered skills are never
 * touched and surface in `SyncResult.conflicts`.
 */

import type { CatalogSourceConfig, SkillCatalog } from "@ratel-ai/sdk";
import { type CatalogSkillWire, projectSkill, skillsEqual } from "./canonical.js";
import { CloudAuthError } from "./errors.js";
import { fetchCatalog } from "./fetch-catalog.js";

export interface SyncResult {
  added: number;
  updated: number;
  removed: number;
  /** Wire ids that collided with host-registered skills and were left untouched. */
  conflicts: string[];
  /** True when the source answered 304 — the replica was revalidated as-is. */
  unchanged: boolean;
}

export interface SkillSyncOptions {
  /** Injectable for tests; defaults to the global `fetch`. */
  fetchImpl?: typeof fetch;
}

// The chain fires at interval ±10% so a fleet of loaders never thunders in step.
const JITTER = 0.1;

export class SkillSync {
  private readonly catalog: SkillCatalog;
  private readonly config: CatalogSourceConfig;
  private readonly fetchImpl?: typeof fetch;

  /** Cached ETag for this loader's fixed `(url, scope)`. */
  private etag?: string;
  private readonly ownedIds = new Set<string>();
  private inFlight?: Promise<SyncResult>;
  private timer?: ReturnType<typeof setTimeout>;
  private intervalMs?: number;
  private authStopped = false;

  private syncedAt?: Date;
  private failures = 0;

  constructor(catalog: SkillCatalog, config: CatalogSourceConfig, options: SkillSyncOptions = {}) {
    this.catalog = catalog;
    this.config = config;
    if (options.fetchImpl !== undefined) this.fetchImpl = options.fetchImpl;
  }

  get lastSyncedAt(): Date | undefined {
    return this.syncedAt;
  }

  get consecutiveFailures(): number {
    return this.failures;
  }

  get ownedCount(): number {
    return this.ownedIds.size;
  }

  /** True once the chain has been shut down — permanently so after an auth error. */
  get stopped(): boolean {
    return this.authStopped;
  }

  /** One conditional fetch + diff apply. Concurrent calls coalesce on the in-flight promise. */
  refresh(): Promise<SyncResult> {
    if (this.inFlight) return this.inFlight;
    this.inFlight = this.doRefresh().finally(() => {
      this.inFlight = undefined;
    });
    return this.inFlight;
  }

  private async doRefresh(): Promise<SyncResult> {
    let fetched: Awaited<ReturnType<typeof fetchCatalog>>;
    try {
      const options = this.fetchImpl
        ? { etag: this.etag, fetchImpl: this.fetchImpl }
        : { etag: this.etag };
      fetched = await fetchCatalog(this.config, options);
    } catch (err) {
      this.failures += 1;
      throw err;
    }

    this.failures = 0;
    this.syncedAt = new Date();
    if (!fetched.changed) {
      return { added: 0, updated: 0, removed: 0, conflicts: [], unchanged: true };
    }

    const result = this.applyDiff(fetched.catalog.skills);
    this.etag = fetched.etag;
    return result;
  }

  private applyDiff(wireSkills: CatalogSkillWire[]): SyncResult {
    const result: SyncResult = {
      added: 0,
      updated: 0,
      removed: 0,
      conflicts: [],
      unchanged: false,
    };
    const wireIds = new Set<string>();

    for (const raw of wireSkills) {
      const skill = projectSkill(raw);
      wireIds.add(skill.id);
      if (this.ownedIds.has(skill.id)) {
        const current = this.catalog.get(skill.id);
        if (current === undefined || !skillsEqual(current, skill)) {
          this.catalog.upsert(skill);
          result.updated += 1;
        }
      } else if (this.catalog.has(skill.id)) {
        result.conflicts.push(skill.id);
      } else {
        this.catalog.upsert(skill);
        this.ownedIds.add(skill.id);
        result.added += 1;
      }
    }

    for (const id of [...this.ownedIds]) {
      if (!wireIds.has(id)) {
        this.catalog.remove(id);
        this.ownedIds.delete(id);
        result.removed += 1;
      }
    }

    return result;
  }

  /**
   * Start the periodic refresh chain: a `setTimeout` chain (never
   * `setInterval`) at `intervalMs` ±10% jitter, `unref()`ed so it can't hold
   * the process open. Transient failures keep the chain alive; an auth error
   * stops it permanently.
   */
  start(intervalMs: number): void {
    if (this.timer !== undefined || this.authStopped) return;
    this.intervalMs = intervalMs;
    this.scheduleTick();
  }

  /** Cancel the pending tick. Idempotent. */
  stop(): void {
    if (this.timer !== undefined) {
      clearTimeout(this.timer);
      this.timer = undefined;
    }
    this.intervalMs = undefined;
  }

  // The tick waits the guaranteed minimum (interval − 10%) first and samples
  // the jitter extra (up to +20% of that point) only when the minimum
  // elapses, so each fire draws fresh randomness.
  private scheduleTick(): void {
    const interval = this.intervalMs;
    if (interval === undefined) return;
    this.setTimer(interval * (1 - JITTER), () => {
      const extra = interval * 2 * JITTER * Math.random();
      // Timers clamp sub-1ms delays to 1ms; a zero extra fires in-tick instead.
      if (extra < 1) {
        void this.tick();
      } else {
        this.setTimer(extra, () => void this.tick());
      }
    });
  }

  private setTimer(delay: number, onFire: () => void): void {
    this.timer = setTimeout(() => {
      this.timer = undefined;
      onFire();
    }, delay);
    this.timer.unref?.();
  }

  private async tick(): Promise<void> {
    try {
      await this.refresh();
    } catch (err) {
      if (err instanceof CloudAuthError) {
        this.authStopped = true;
        this.stop();
        return;
      }
      // Transient (network / 5xx / unexpected): replica stays live, chain continues.
    }
    this.scheduleTick();
  }
}
